probe/plugins/registry.go

```diff
@@ -0,0 +1,293 @@
+		pluginReport.Plugins = pluginReport.Plugins.Add(plugin.PluginSpec)
+		rpt = rpt.Merge(pluginReport)
+	})
+	return rpt, nil
+}
+
+// Close shuts down the registry. It can still be used after this, but will be
+// out of date.
+func (r *Registry) Close() {
+	r.cancel()
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	for _, plugin := range r.pluginsBySocket {
+		plugin.Close()
+	}
+}
+
+// Plugin is the implementation of a plugin. It is responsible for doing the
+// plugin handshake, gathering reports, etc.
+type Plugin struct {
+	xfer.PluginSpec
+	context context.Context
+	socket  string
+	client  *http.Client
+	cancel  context.CancelFunc
+	backoff backoff.Interface
+}
+
+// NewPlugin loads and initializes a new plugin. If client is nil,
+// http.DefaultClient will be used.
+func NewPlugin(ctx context.Context, socket string, client *http.Client, expectedAPIVersion string, handshakeMetadata map[string]string) *Plugin {
+	params := url.Values{}
+	for k, v := range handshakeMetadata {
+		params.Add(k, v)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	p := &Plugin{context: ctx, socket: socket, client: client, cancel: cancel}
+	f := p.handshake(ctx, expectedAPIVersion, params)
+	f() // try the first time synchronously
+	p.backoff = backoff.New(f, "plugin handshake")
+	go p.backoff.Start()
+	return p
+}
+
+type handshakeResponse struct {
+	Name        string   `json:"name"`
+	Description string   `json:"description,omitempty"`
+	Interfaces  []string `json:"interfaces"`
+	APIVersion  string   `json:"api_version,omitempty"`
+}
+
+// handshake tries the handshake with this plugin.
+func (p *Plugin) handshake(ctx context.Context, expectedAPIVersion string, params url.Values) func() (bool, error) {
+	return func() (ok bool, err error) {
+		defer func() { p.setStatus(err) }()
+		var resp handshakeResponse
+		if err := p.get("/", params, &resp); err != nil {
+			return err == context.Canceled, fmt.Errorf("plugins: error loading plugin %s: %v", p.socket, err)
+		}
+
+		if resp.Name == "" {
+			return false, fmt.Errorf("plugins: error loading plugin %s: plugin did not provide a name", p.socket)
+		}
+		if resp.APIVersion != expectedAPIVersion {
+			return false, fmt.Errorf("plugins: error loading plugin %s: plugin did not provide correct API version: expected %q, got %q", p.socket, expectedAPIVersion, resp.APIVersion)
+		}
+		p.ID, p.Label = resp.Name, resp.Name
+		p.Description = resp.Description
+		p.Interfaces = resp.Interfaces
+		log.Infof("plugins: loaded plugin %s: %s", p.ID, strings.Join(p.Interfaces, ", "))
+		return true, nil
+	}
+}
+
+// Report gets the latest report from the plugin
+func (p *Plugin) Report() (report.Report, error) {
+	result := report.MakeReport()
+	err := p.get("/report", nil, &result)
+	p.setStatus(err)
+	return result, err
+}
+
+func (p *Plugin) setStatus(err error) {
+	if err == nil {
+		p.Status = "ok"
+	} else {
+		p.Status = fmt.Sprintf("error: %v", err)
+	}
+}
+
+// TODO(paulbellamy): better error handling on wrong status codes
```

Comment:
I confirm this is needed right now for debugging plugins (my python plugin was originally failing with 404 errors and scope was logging cryptic "unexpected EOF" because it was trying to decode the body).